clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -402,27 +403,41 @@ public void refreshCommittedOffsetsIfNeeded() {
         }
     }
 
-    @Override
-    public void close() {
+    public void close(long timeoutMs, long requestTimeoutMs) {
         // we do not need to re-enable wakeups since we are closing already
         client.disableWakeups();
+
+        long now = time.milliseconds();
+        timeoutMs = Math.min(timeoutMs, requestTimeoutMs);
+        long endTimeMs = now + timeoutMs;
         try {
-            maybeAutoCommitOffsetsSync();
-
-            Node coordinator;
-            long endTimeMs = time.milliseconds() + CLOSE_TIMEOUT_MS;
-            while ((coordinator = coordinator()) != null && client.pendingRequestCount(coordinator) > 0) {
-                long remainingTimeMs = endTimeMs - time.milliseconds();
-                if (remainingTimeMs > 0)
-                    client.poll(remainingTimeMs);
-                else {
-                    log.warn("Close timed out with {} pending requests to coordinator, terminating client connections for group {}.", client.pendingRequestCount(coordinator), groupId);
-                    break;
-                }
+            maybeAutoCommitOffsetsSync(timeoutMs);
+            now = time.milliseconds();
+            if (pendingAsyncCommits.get() > 0) {
```

Comment:
If there are pending async commits, then the coordinator must be known (because we explicitly fail all requests to the coordinator in `coordinatorDead`). However, I think there is some value in calling `ensureCoordinatorReady` prior to invoking `maybeAutoCommitOffsetsSync` and also prior to sending the LeaveGroup.